File: my-folder/0110-balanced-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    
    var dfs func(node *TreeNode) (bool, int)

    dfs = func(node *TreeNode) (bool, int) {
        if node == nil {
            return true, 0
        }
        ln, lv := dfs(node.Left)
        rn, rv := dfs(node.Right)

        if !ln || !rn {
            return false, 0
        }

        diff := abs(lv - rv)
        if diff <= 1 {
            return true, 1 + max(lv, rv)
        } else {
            return false, 1 + max(lv, rv)
        }
    }

    b, _ := dfs(root)

    return b
}

func abs(v int) int {
    if v < 0 {
        return -v
    }
    return v
}

File: my-folder/0110-balanced-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {

    let checkHeight = function(root: TreeNode | null): number {
        if (root == null) {
            return 0
        }

        let left = checkHeight(root.left)
        let right = checkHeight(root.right)

        if (left === -1 || right === -1) {
            return -1
        }
        if (Math.abs(left - right) > 1) {
            return -1
        }

        return 1 + Math.max(left, right)
    }

    return checkHeight(root) !== -1
};
